File: Java-Approvals/src/main/java/com/gildedrose/GildedRose.java
package com.gildedrose;

class GildedRose {
    public static final int MAX_QUALITY = 50;
    public static final int MIN_QUALITY = 0;
    Item[] items;

    public GildedRose(Item[] items) {
        this.items = items;
    }

    public void endDay() {
        for (Item item : items) {
            updateSellIn(item);
            updateQuality(item);
        }
    }

    private static void updateQuality(Item item) {
        if (isSulfuras(item)) {
        } else if (isBackstagePass(item)) {
            increaseQuality(item, 1);
            if (expiresInDays(item, 10)) {
                increaseQuality(item, 1);
            }
            if (expiresInDays(item, 5)) {
                increaseQuality(item, 1);
            }
            if (isExpired(item)) {
                increaseQuality(item, -item.quality);
            }
        } else if (isAgedBrie(item)) {
            increaseQuality(item, 1);
            if (isExpired(item)) {
                increaseQuality(item, 1);
            }
        } else {
            increaseQuality(item, -1);
            if (isExpired(item)) {
                increaseQuality(item, -1);
            }
        }
    }

    private static boolean expiresInDays(Item item, int days) {
        return item.sellIn < days;
    }

    private static boolean isExpired(Item item) {
        return expiresInDays(item, 0);
    }

    private static void updateSellIn(Item item) {
        if (isSulfuras(item)) {
        } else {
            item.sellIn = item.sellIn - 1;
        }
    }

    private static void increaseQuality(Item item, int number) {
        item.quality = limitQualityRange(item.quality + number);
    }

    private static boolean isSulfuras(Item item) {
        return item.name.equals("Sulfuras, Hand of Ragnaros");
    }

    private static boolean isBackstagePass(Item item) {
        return item.name.equals("Backstage passes to a TAFKAL80ETC concert");
    }

    private static boolean isAgedBrie(Item item) {
        return item.name.equals("Aged Brie");
    }

    private static int limitQualityRange(int value) {
        return Math.min(Math.max(value, MIN_QUALITY), MAX_QUALITY);
    }
}

File: Kotlin/src/main/kotlin/com/gildedrose/GildedRose.kt
package com.gildedrose

const val AGED_BRIE = "Aged Brie"
const val BACKSTAGE_PASSES = "Backstage passes to a TAFKAL80ETC concert"
const val SULFURAS = "Sulfuras, Hand of Ragnaros"
const val CONJURED = "Conjured Mana Cake"

private const val MIN_QUALITY = 0
private const val MAX_QUALITY = 50
private const val LEGENDARY_QUALITY = 80

class GildedRose(val items: List<Item>) {

    fun updateQuality() {
        items.forEach { item ->
            updaterFor(item).update(item)
        }
    }
}

private fun updaterFor(item: Item): ItemUpdater = when (item.name) {
    SULFURAS -> LegendaryItemUpdater
    AGED_BRIE -> AgedBrieUpdater
    BACKSTAGE_PASSES -> BackstagePassUpdater
    CONJURED -> ConjuredItemUpdater
    else -> NormalItemUpdater
}

private interface ItemUpdater {
    fun update(item: Item)
}

private object LegendaryItemUpdater : ItemUpdater {
    override fun update(item: Item) {
        item.quality = LEGENDARY_QUALITY
    }
}

private object AgedBrieUpdater : ItemUpdater {
    override fun update(item: Item) {
        item.sellIn--
        val increase = if (item.expired) 2 else 1
        item.quality = (item.quality + increase).coerceIn(MIN_QUALITY, MAX_QUALITY)
    }
}

private object BackstagePassUpdater : ItemUpdater {
    override fun update(item: Item) {
        item.sellIn--
        item.quality = when {
            item.expired -> 0
            item.sellIn < 5 -> item.quality + 3
            item.sellIn < 10 -> item.quality + 2
            else -> item.quality + 1
        }.coerceIn(MIN_QUALITY, MAX_QUALITY)
    }
}

private object ConjuredItemUpdater : ItemUpdater {
    override fun update(item: Item) {
        item.sellIn--
        val degradation = if (item.expired) 4 else 2
        item.quality = (item.quality - degradation).coerceIn(MIN_QUALITY, MAX_QUALITY)
    }
}

private object NormalItemUpdater : ItemUpdater {
    override fun update(item: Item) {
        item.sellIn--
        val degradation = if (item.expired) 2 else 1
        item.quality = (item.quality - degradation).coerceIn(MIN_QUALITY, MAX_QUALITY)
    }
}

private val Item.expired get() = sellIn < 0
